File: clients/go/consensus/connect_block_inmem.go
package consensus

import "math/big"

// InMemoryChainState is a minimal, non-persistent chainstate container intended for
// conformance tests and audit/repro tooling.
//
// It is NOT a full node database: it provides deterministic inputs for stateful checks
// (fee computation, coinbase bound, maturity rules) without any disk persistence.
type InMemoryChainState struct {
	Utxos            map[Outpoint]UtxoEntry
	AlreadyGenerated *big.Int // already_generated(h): subsidy-only, excluding fees
}

type ConnectBlockBasicSummary struct {
	SumFees            Uint128
	AlreadyGenerated   Uint128
	AlreadyGeneratedN1 Uint128
	UtxoCount          uint64
	PostStateDigest    [32]byte

	// SigTaskCount is the number of signature verification tasks dispatched
	// to the worker pool during parallel validation. Zero for sequential path.
	SigTaskCount uint64
	// WorkerPanics is the number of panics recovered in worker goroutines
	// during parallel validation. Zero for sequential path.
	WorkerPanics uint64
}

type connectBlockBasicInMemorySuiteContext struct {
	BlockBytes       []byte
	ExpectedPrevHash *[32]byte
	ExpectedTarget   *[32]byte
	BlockHeight      uint64
	PrevTimestamps   []uint64
	State            *InMemoryChainState
	ChainID          [32]byte
	Rotation         RotationProvider
	Registry         *SuiteRegistry
}

type connectBlockInMemoryValidationContext struct {
	chainID  [32]byte
	rotation RotationProvider
	registry *SuiteRegistry
}

// ConnectBlockBasicInMemoryAtHeight connects a block against an in-memory UTXO snapshot and an
// in-memory subsidy counter, and enforces the coinbase subsidy/value bound (CANONICAL §19.2)
// using locally computed fees.
//
// This closes the "partial" nature of F-05: sum_fees is computed internally as
// Σ(sum_in - sum_out) over all non-coinbase transactions in the block.
//
// Persistence (writing chainstate to disk) is intentionally out of scope here.
func ConnectBlockBasicInMemoryAtHeight(
	blockBytes []byte,
	expectedPrevHash *[32]byte,
	expectedTarget *[32]byte,
	blockHeight uint64,
	prevTimestamps []uint64,
	state *InMemoryChainState,
	chainID [32]byte,
) (*ConnectBlockBasicSummary, error) {
	return ConnectBlockBasicInMemoryAtHeightAndSuiteContext(
		blockBytes,
		expectedPrevHash,
		expectedTarget,
		blockHeight,
		prevTimestamps,
		state,
		chainID,
		nil, /*rotation*/
		nil, /*registry*/
	)
}

// #lizard forgive
func ConnectBlockBasicInMemoryAtHeightAndSuiteContext(
	blockBytes []byte,
	expectedPrevHash *[32]byte,
	expectedTarget *[32]byte,
	blockHeight uint64,
	prevTimestamps []uint64,
	state *InMemoryChainState,
	chainID [32]byte,
	rotation RotationProvider,
	registry *SuiteRegistry,
) (*ConnectBlockBasicSummary, error) {
	return connectBlockBasicInMemoryAtHeightAndSuiteContext(connectBlockBasicInMemorySuiteContext{
		BlockBytes:       blockBytes,
		ExpectedPrevHash: expectedPrevHash,
		ExpectedTarget:   expectedTarget,
		BlockHeight:      blockHeight,
		PrevTimestamps:   prevTimestamps,
		State:            state,
		ChainID:          chainID,
		Rotation:         rotation,
		Registry:         registry,
	})
}

func connectBlockBasicInMemoryAtHeightAndSuiteContext(
	input connectBlockBasicInMemorySuiteContext,
) (*ConnectBlockBasicSummary, error) {
	alreadyGenerated, alreadyGeneratedU128, err := preflightInMemoryChainState(input.State)
	if err != nil {
		return nil, err
	}

	pb, err := parseInMemoryConnectBlock(input)
	if err != nil {
		return nil, err
	}

	blockMTP, err := inMemoryConnectBlockMTP(input.BlockHeight, input.PrevTimestamps, pb.Header.Timestamp)
	if err != nil {
		return nil, err
	}

	validation := connectBlockInMemoryValidationContext{
		chainID:  input.ChainID,
		rotation: input.Rotation,
		registry: input.Registry,
	}
	workUtxos := cloneUtxoSet(input.State.Utxos)
	workUtxos, sumFees, err := applyInMemorySequentialConnect(
		pb,
		workUtxos,
		input.BlockHeight,
		blockMTP,
		validation,
	)
	if err != nil {
		return nil, err
	}

	if err := validateCoinbaseValueBound(pb, input.BlockHeight, alreadyGenerated, sumFees); err != nil {
		return nil, err
	}

	alreadyGeneratedN1, err := checkedAdvanceAlreadyGenerated(input.BlockHeight, alreadyGeneratedU128, alreadyGenerated)
	if err != nil {
		return nil, err
	}
	return commitInMemoryConnectSummary(input.State, workUtxos, alreadyGeneratedU128, alreadyGeneratedN1, sumFees, 0, 0), nil
}

// applyInMemorySequentialConnect is the sequential production continuation
// after parsed-block preflight. It makes the coinbase outputs visible to every
// non-coinbase transaction while retaining all effects in the caller's local
// working UTXO set until final commit.
func applyInMemorySequentialConnect(
	pb *ParsedBlock,
	workUtxos map[Outpoint]UtxoEntry,
	blockHeight uint64,
	blockMTP uint64,
	validation connectBlockInMemoryValidationContext,
) (map[Outpoint]UtxoEntry, Uint128, error) {
	if err := applyInMemoryCoinbaseOutputs(
		pb,
		workUtxos,
		blockHeight,
		validation.chainID,
		validation.rotation,
	); err != nil {
		return nil, Uint128{}, err
	}
	return applyInMemoryNonCoinbaseTxs(
		pb,
		workUtxos,
		blockHeight,
		blockMTP,
		validation,
	)
}

func preflightInMemoryChainState(state *InMemoryChainState) (*big.Int, Uint128, error) {
	if state == nil {
		return nil, Uint128{}, txerr(BLOCK_ERR_PARSE, "nil chainstate")
	}
	alreadyGenerated := new(big.Int)
	if state.AlreadyGenerated != nil {
		alreadyGenerated.Set(state.AlreadyGenerated)
	}
	if alreadyGenerated.Sign() < 0 {
		return nil, Uint128{}, txerr(BLOCK_ERR_PARSE, "already_generated must be unsigned")
	}
	if alreadyGenerated.BitLen() > 128 {
		return nil, Uint128{}, txerr(BLOCK_ERR_PARSE, "already_generated overflow")
	}
	return alreadyGenerated, uint128FromBigInt(alreadyGenerated), nil
}

func parseInMemoryConnectBlock(input connectBlockBasicInMemorySuiteContext) (*ParsedBlock, error) {
	pb, err := parseAndValidateBlockBasicForConnectWithContextAtHeight(
		input.BlockBytes,
		input.ExpectedPrevHash,
		input.ExpectedTarget,
		input.BlockHeight,
		input.PrevTimestamps,
		input.ChainID,
		input.Rotation,
	)
	if err != nil {
		return nil, err
	}
	if pb == nil || len(pb.Txs) == 0 || len(pb.Txids) != len(pb.Txs) {
		return nil, txerr(BLOCK_ERR_PARSE, "invalid parsed block")
	}
	return pb, nil
}

func inMemoryConnectBlockMTP(blockHeight uint64, prevTimestamps []uint64, headerTimestamp uint64) (uint64, error) {
	median, ok, err := medianTimePast(blockHeight, prevTimestamps)
	if err != nil {
		return 0, err
	}
	if ok {
		return median, nil
	}
	return headerTimestamp, nil
}

func applyInMemoryNonCoinbaseTxs(
	pb *ParsedBlock,
	workUtxos map[Outpoint]UtxoEntry,
	blockHeight uint64,
	blockMTP uint64,
	validation connectBlockInMemoryValidationContext,
) (map[Outpoint]UtxoEntry, Uint128, error) {
	var sumFees Uint128
	seenNonces := make(map[uint64]struct{}, len(pb.Txs))
	for i := 1; i < len(pb.Txs); i++ {
		if err := validateNonCoinbaseBlockTx(pb.Txs[i], seenNonces); err != nil {
			return nil, Uint128{}, err
		}
		nextUtxos, fee, err := applyNonCoinbaseTxBasicWork(nonCoinbaseApplyWorkInput{
			tx:       pb.Txs[i],
			txid:     pb.Txids[i],
			utxoSet:  workUtxos,
			height:   blockHeight,
			blockMTP: blockMTP,
			chainID:  validation.chainID,
			rotation: validation.rotation,
			registry: validation.registry,
		})
		if err != nil {
			return nil, Uint128{}, err
		}
		workUtxos = nextUtxos
		next, ok := sumFees.CheckedAdd(fee)
		if !ok {
			return nil, Uint128{}, txerr(BLOCK_ERR_PARSE, "sum_fees overflow")
		}
		sumFees = next
	}
	return workUtxos, sumFees, nil
}

// applyInMemoryCoinbaseOutputs validates and applies coinbase outputs in vout
// order before transaction index one. Every output passes the coinbase-only
// VAULT guard and the existing creation validator before its spendable UTXO is
// inserted. The same-CMR cap runs only after every output is processed.
func applyInMemoryCoinbaseOutputs(
	pb *ParsedBlock,
	workUtxos map[Outpoint]UtxoEntry,
	blockHeight uint64,
	chainID [32]byte,
	rotation RotationProvider,
) error {
	coinbase := pb.Txs[0]
	coinbaseTxid := pb.Txids[0]
	if rotation == nil {
		rotation = DefaultRotationProvider{}
	}
	simplicityDeployment := simplicityDeploymentFromRotation(rotation)
	simplicityOutputCMRs := make([][32]byte, 0, len(coinbase.Outputs))
	for i, out := range coinbase.Outputs {
		if err := validateCoinbaseApplyOutput(out); err != nil {
			return err
		}
		programCMR, isCoreSimplicity, err := validateTxOutputCovenantGenesis(
			coinbase.TxKind,
			out,
			chainID,
			blockHeight,
			rotation,
			simplicityDeployment,
		)
		if err != nil {
			return err
		}
		if isCoreSimplicity {
			simplicityOutputCMRs = append(simplicityOutputCMRs, programCMR)
		}
		if isNonSpendableInputCovenant(out.CovenantType) {
			continue
		}
		vout, err := coinbaseOutputVout(i)
		if err != nil {
			return err
		}
		op := Outpoint{Txid: coinbaseTxid, Vout: vout}
		workUtxos[op] = UtxoEntry{
			Value:             out.Value,
			CovenantType:      out.CovenantType,
			CovenantData:      append([]byte(nil), out.CovenantData...),
			CreationHeight:    blockHeight,
			CreatedByCoinbase: true,
		}
	}
	return validateSimplicityOutputGroupCap(simplicityOutputCMRs)
}

func coinbaseOutputVout(index int) (uint32, error) {
	if index < 0 || uint64(index) > uint64(^uint32(0)) {
		return 0, txerr(BLOCK_ERR_PARSE, "coinbase output index exceeds u32")
	}
	return uint32(index), nil
}

func checkedAdvanceAlreadyGenerated(blockHeight uint64, alreadyGenerated Uint128, alreadyGeneratedBig *big.Int) (Uint128, error) {
	if blockHeight == 0 {
		return alreadyGenerated, nil
	}
	alreadyGeneratedN1, ok := alreadyGenerated.CheckedAdd(Uint128FromU64(BlockSubsidyBig(blockHeight, alreadyGeneratedBig)))
	if !ok {
		return Uint128{}, txerr(BLOCK_ERR_PARSE, "already_generated overflow")
	}
	return alreadyGeneratedN1, nil
}

func commitInMemoryConnectSummary(
	state *InMemoryChainState,
	workUtxos map[Outpoint]UtxoEntry,
	alreadyGenerated Uint128,
	alreadyGeneratedN1 Uint128,
	sumFees Uint128,
	sigTaskCount uint64,
	workerPanics uint64,
) *ConnectBlockBasicSummary {
	summary := &ConnectBlockBasicSummary{
		SumFees:            sumFees,
		AlreadyGenerated:   alreadyGenerated,
		AlreadyGeneratedN1: alreadyGeneratedN1,
		UtxoCount:          uint64(len(workUtxos)),
		PostStateDigest:    UtxoSetHash(workUtxos),
		SigTaskCount:       sigTaskCount,
		WorkerPanics:       workerPanics,
	}
	state.Utxos = workUtxos
	state.AlreadyGenerated = alreadyGeneratedN1.Big()
	return summary
}

func uint128FromBigInt(v *big.Int) Uint128 {
	hi := new(big.Int).Rsh(new(big.Int).Set(v), 64)
	return Uint128{Lo: v.Uint64(), Hi: hi.Uint64()}
}

File: clients/rust/crates/rubin-consensus/src/connect_block_inmem.rs
use std::collections::HashMap;

use sha3::{Digest, Sha3_256};

use crate::block_basic::{
    median_time_past, parse_block_bytes, validate_coinbase_apply_outputs,
    validate_coinbase_value_bound, validate_parsed_block_basic_with_context_at_height, ParsedBlock,
};
use crate::compactsize::encode_compact_size;
use crate::constants::{COV_TYPE_ANCHOR, COV_TYPE_DA_COMMIT};
use crate::core_ext::{CoreExtDeploymentProfiles, CoreExtProfiles};
use crate::error::{ErrorCode, TxError};
use crate::sig_queue::SigCheckQueue;
use crate::subsidy::block_subsidy;
use crate::suite_registry::{RotationProvider, SuiteRegistry};
use crate::utxo_basic::{
    apply_non_coinbase_tx_basic_update_with_mtp_and_core_ext_profiles_and_suite_context,
    apply_non_coinbase_tx_basic_update_with_mtp_and_core_ext_profiles_and_suite_context_queued_sigchecks,
    Outpoint, UtxoEntry,
};

const UTXO_SET_HASH_DST: &[u8] = b"RUBINv1-utxo-set-hash/";

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct InMemoryChainState {
    pub utxos: HashMap<Outpoint, UtxoEntry>,
    /// already_generated(h): subsidy-only (excluding fees).
    pub already_generated: u128,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ConnectBlockBasicSummary {
    pub sum_fees: u64,
    pub already_generated: u128,
    pub already_generated_n1: u128,
    pub utxo_count: u64,
    /// Post-state UTXO set digest (SHA3-256) for parity checks.
    pub post_state_digest: [u8; 32],
    /// Number of queued signature-verification tasks. Zero for sequential path.
    pub sig_task_count: u64,
    /// Number of recovered worker panics. Zero on successful validation.
    pub worker_panics: u64,
}

struct ConnectBlockContext<'a> {
    expected_prev_hash: Option<[u8; 32]>,
    expected_target: Option<[u8; 32]>,
    block_height: u64,
    prev_timestamps: Option<&'a [u64]>,
    chain_id: [u8; 32],
    core_ext_deployments: &'a CoreExtDeploymentProfiles,
    rotation: Option<&'a dyn RotationProvider>,
    registry: Option<&'a SuiteRegistry>,
}

struct PreparedConnectBlock {
    pb: ParsedBlock,
    block_height: u64,
    already_generated: u128,
    block_mtp: u64,
    core_ext_profiles: CoreExtProfiles,
}

/// ConnectBlockBasicInMemoryAtHeight connects a block against an in-memory chainstate and enforces
/// the coinbase subsidy/value bound using locally computed fees.
///
/// This intentionally does not provide any on-disk persistence.
pub fn connect_block_basic_in_memory_at_height(
    block_bytes: &[u8],
    expected_prev_hash: Option<[u8; 32]>,
    expected_target: Option<[u8; 32]>,
    block_height: u64,
    prev_timestamps: Option<&[u64]>,
    state: &mut InMemoryChainState,
    chain_id: [u8; 32],
) -> Result<ConnectBlockBasicSummary, TxError> {
    connect_block_basic_in_memory_at_height_and_core_ext_deployments(
        block_bytes,
        expected_prev_hash,
        expected_target,
        block_height,
        prev_timestamps,
        state,
        chain_id,
        &CoreExtDeploymentProfiles::empty(),
    )
}

#[allow(clippy::too_many_arguments)]
pub fn connect_block_basic_in_memory_at_height_and_core_ext_deployments(
    block_bytes: &[u8],
    expected_prev_hash: Option<[u8; 32]>,
    expected_target: Option<[u8; 32]>,
    block_height: u64,
    prev_timestamps: Option<&[u64]>,
    state: &mut InMemoryChainState,
    chain_id: [u8; 32],
    core_ext_deployments: &CoreExtDeploymentProfiles,
) -> Result<ConnectBlockBasicSummary, TxError> {
    connect_block_basic_in_memory_at_height_and_core_ext_deployments_with_suite_context(
        block_bytes,
        expected_prev_hash,
        expected_target,
        block_height,
        prev_timestamps,
        state,
        chain_id,
        core_ext_deployments,
        None,
        None,
    )
}

#[allow(clippy::too_many_arguments)]
pub fn connect_block_basic_in_memory_at_height_and_core_ext_deployments_with_suite_context(
    block_bytes: &[u8],
    expected_prev_hash: Option<[u8; 32]>,
    expected_target: Option<[u8; 32]>,
    block_height: u64,
    prev_timestamps: Option<&[u64]>,
    state: &mut InMemoryChainState,
    chain_id: [u8; 32],
    core_ext_deployments: &CoreExtDeploymentProfiles,
    rotation: Option<&dyn RotationProvider>,
    registry: Option<&SuiteRegistry>,
) -> Result<ConnectBlockBasicSummary, TxError> {
    let ctx = ConnectBlockContext {
        expected_prev_hash,
        expected_target,
        block_height,
        prev_timestamps,
        chain_id,
        core_ext_deployments,
        rotation,
        registry,
    };
    connect_block_basic_in_memory_with_context(block_bytes, state, &ctx)
}

/// Go-style block-level deferred signature orchestration. Structural and
/// state-mutation checks stay sequential; only expensive native signature
/// verification is deferred and flushed once per block.
#[allow(clippy::too_many_arguments)]
pub fn connect_block_parallel_sig_verify(
    block_bytes: &[u8],
    expected_prev_hash: Option<[u8; 32]>,
    expected_target: Option<[u8; 32]>,
    block_height: u64,
    prev_timestamps: Option<&[u64]>,
    state: &mut InMemoryChainState,
    chain_id: [u8; 32],
    workers: usize,
) -> Result<ConnectBlockBasicSummary, TxError> {
    connect_block_parallel_sig_verify_and_core_ext_deployments(
        block_bytes,
        expected_prev_hash,
        expected_target,
        block_height,
        prev_timestamps,
        state,
        chain_id,
        &CoreExtDeploymentProfiles::empty(),
        workers,
    )
}

#[allow(clippy::too_many_arguments)]
pub fn connect_block_parallel_sig_verify_and_core_ext_deployments(
    block_bytes: &[u8],
    expected_prev_hash: Option<[u8; 32]>,
    expected_target: Option<[u8; 32]>,
    block_height: u64,
    prev_timestamps: Option<&[u64]>,
    state: &mut InMemoryChainState,
    chain_id: [u8; 32],
    core_ext_deployments: &CoreExtDeploymentProfiles,
    workers: usize,
) -> Result<ConnectBlockBasicSummary, TxError> {
    connect_block_parallel_sig_verify_and_core_ext_deployments_with_suite_context(
        block_bytes,
        expected_prev_hash,
        expected_target,
        block_height,
        prev_timestamps,
        state,
        chain_id,
        core_ext_deployments,
        None,
        None,
        workers,
    )
}

#[allow(clippy::too_many_arguments)]
pub fn connect_block_parallel_sig_verify_and_core_ext_deployments_with_suite_context(
    block_bytes: &[u8],
    expected_prev_hash: Option<[u8; 32]>,
    expected_target: Option<[u8; 32]>,
    block_height: u64,
    prev_timestamps: Option<&[u64]>,
    state: &mut InMemoryChainState,
    chain_id: [u8; 32],
    core_ext_deployments: &CoreExtDeploymentProfiles,
    rotation: Option<&dyn RotationProvider>,
    registry: Option<&SuiteRegistry>,
    workers: usize,
) -> Result<ConnectBlockBasicSummary, TxError> {
    let ctx = ConnectBlockContext {
        expected_prev_hash,
        expected_target,
        block_height,
        prev_timestamps,
        chain_id,
        core_ext_deployments,
        rotation,
        registry,
    };
    connect_block_parallel_sig_verify_with_context(block_bytes, state, &ctx, workers)
}

fn prepare_connect_block(
    block_bytes: &[u8],
    already_generated: u128,
    ctx: &ConnectBlockContext<'_>,
) -> Result<PreparedConnectBlock, TxError> {
    // G.9: parse once and validate against the parsed block.
    let pb = parse_block_bytes(block_bytes)?;
    validate_parsed_block_basic_with_context_at_height(
        &pb,
        ctx.expected_prev_hash,
        ctx.expected_target,
        ctx.block_height,
        ctx.prev_timestamps,
    )?;
    if pb.txs.is_empty() || pb.txids.len() != pb.txs.len() {
        return Err(TxError::new(
            ErrorCode::BlockErrParse,
            "invalid parsed block",
        ));
    }

    // 0x0102 (CORE_EXT) is unassigned with no activation path, so the retired
    // deployment set must not influence consensus/admission for ordinary
    // (e.g. P2PK) traffic. Do NOT call `active_profiles_at_height` here: it
    // validates the deployment set and would reject otherwise-valid blocks if a
    // node were started with a stale/invalid `core_ext_deployments` entry. Pass
    // empty active profiles instead, mirroring Go, which ignores the parameter.
    // The parameter is retained on the public signatures for node/CLI API
    // compatibility; full removal is tracked under CORE_EXT retirement.
    let _ = ctx.core_ext_deployments;
    Ok(PreparedConnectBlock {
        block_mtp: median_time_past(ctx.block_height, ctx.prev_timestamps)?
            .unwrap_or(pb.header.timestamp),
        core_ext_profiles: CoreExtProfiles::default(),
        pb,
        block_height: ctx.block_height,
        already_generated,
    })
}

fn connect_block_basic_in_memory_with_context(
    block_bytes: &[u8],
    state: &mut InMemoryChainState,
    ctx: &ConnectBlockContext<'_>,
) -> Result<ConnectBlockBasicSummary, TxError> {
    let prepared = prepare_connect_block(block_bytes, state.already_generated, ctx)?;
    let (work_utxos, sum_fees) = apply_non_coinbase_txs_sequential(&prepared, &state.utxos, ctx)?;
    finalize_connected_block(state, &prepared, work_utxos, sum_fees, 0)
}

fn apply_non_coinbase_txs_sequential(
    prepared: &PreparedConnectBlock,
    state_utxos: &HashMap<Outpoint, UtxoEntry>,
    ctx: &ConnectBlockContext<'_>,
) -> Result<(HashMap<Outpoint, UtxoEntry>, u64), TxError> {
    let mut work_utxos = None;
    let mut sum_fees: u64 = 0;
    for i in 1..prepared.pb.txs.len() {
        let base_utxos = work_utxos.as_ref().unwrap_or(state_utxos);
        let (next_utxos, summary) =
            apply_non_coinbase_tx_basic_update_with_mtp_and_core_ext_profiles_and_suite_context(
                &prepared.pb.txs[i],
                prepared.pb.txids[i],
                base_utxos,
                prepared.block_height,
                prepared.pb.header.timestamp,
                prepared.block_mtp,
                ctx.chain_id,
                &prepared.core_ext_profiles,
                ctx.rotation,
                ctx.registry,
            )?;
        work_utxos = Some(next_utxos);
        sum_fees = add_block_fee(sum_fees, summary.fee)?;
    }

    Ok((work_utxos.unwrap_or_else(|| state_utxos.clone()), sum_fees))
}

fn connect_block_parallel_sig_verify_with_context(
    block_bytes: &[u8],
    state: &mut InMemoryChainState,
    ctx: &ConnectBlockContext<'_>,
    workers: usize,
) -> Result<ConnectBlockBasicSummary, TxError> {
    let prepared = prepare_connect_block(block_bytes, state.already_generated, ctx)?;
    let (work_utxos, sum_fees, sig_task_count) =
        apply_non_coinbase_txs_parallel(&prepared, &state.utxos, ctx, workers)?;
    finalize_connected_block(state, &prepared, work_utxos, sum_fees, sig_task_count)
}

fn apply_non_coinbase_txs_parallel(
    prepared: &PreparedConnectBlock,
    state_utxos: &HashMap<Outpoint, UtxoEntry>,
    ctx: &ConnectBlockContext<'_>,
    workers: usize,
) -> Result<(HashMap<Outpoint, UtxoEntry>, u64, u64), TxError> {
    let mut work_utxos = state_utxos.clone();
    let mut sig_queue = match ctx.registry {
        Some(registry) => SigCheckQueue::new(workers).with_registry(registry),
        None => SigCheckQueue::new(workers),
    };

    let mut sum_fees: u64 = 0;
    for i in 1..prepared.pb.txs.len() {
        let (next_utxos, summary) =
            apply_non_coinbase_tx_basic_update_with_mtp_and_core_ext_profiles_and_suite_context_queued_sigchecks(
                &prepared.pb.txs[i],
                prepared.pb.txids[i],
                &work_utxos,
                prepared.block_height,
                prepared.pb.header.timestamp,
                prepared.block_mtp,
                ctx.chain_id,
                &prepared.core_ext_profiles,
                ctx.rotation,
                ctx.registry,
                &mut sig_queue,
            )?;
        work_utxos = next_utxos;
        sum_fees = add_block_fee(sum_fees, summary.fee)?;
    }

    let sig_task_count = sig_queue.len() as u64;
    sig_queue.flush()?;

    Ok((work_utxos, sum_fees, sig_task_count))
}

fn add_block_fee(sum_fees: u64, fee: u64) -> Result<u64, TxError> {
    sum_fees
        .checked_add(fee)
        .ok_or_else(|| TxError::new(ErrorCode::BlockErrParse, "sum_fees overflow"))
}

fn finalize_connected_block(
    state: &mut InMemoryChainState,
    prepared: &PreparedConnectBlock,
    mut work_utxos: HashMap<Outpoint, UtxoEntry>,
    sum_fees: u64,
    sig_task_count: u64,
) -> Result<ConnectBlockBasicSummary, TxError> {
    validate_coinbase_value_bound(
        &prepared.pb,
        prepared.block_height,
        prepared.already_generated,
        sum_fees,
    )?;
    validate_coinbase_apply_outputs(&prepared.pb.txs[0])?;
    add_coinbase_outputs(&mut work_utxos, prepared);
    let already_generated_n1 =
        already_generated_after_block(prepared.block_height, prepared.already_generated)?;

    state.utxos = work_utxos;
    if prepared.block_height != 0 {
        state.already_generated = already_generated_n1;
    }

    let post_state_digest = utxo_set_hash(&state.utxos);

    Ok(ConnectBlockBasicSummary {
        sum_fees,
        already_generated: prepared.already_generated,
        already_generated_n1,
        utxo_count: state.utxos.len() as u64,
        post_state_digest,
        sig_task_count,
        worker_panics: 0,
    })
}

fn add_coinbase_outputs(
    work_utxos: &mut HashMap<Outpoint, UtxoEntry>,
    prepared: &PreparedConnectBlock,
) {
    let coinbase_txid = prepared.pb.txids[0];
    for (i, out) in prepared.pb.txs[0].outputs.iter().enumerate() {
        if out.covenant_type == COV_TYPE_ANCHOR || out.covenant_type == COV_TYPE_DA_COMMIT {
            continue;
        }
        work_utxos.insert(
            Outpoint {
                txid: coinbase_txid,
                vout: i as u32,
            },
            UtxoEntry {
                value: out.value,
                covenant_type: out.covenant_type,
                covenant_data: out.covenant_data.clone(),
                creation_height: prepared.block_height,
                created_by_coinbase: true,
            },
        );
    }
}

fn already_generated_after_block(
    block_height: u64,
    already_generated: u128,
) -> Result<u128, TxError> {
    if block_height != 0 {
        let subsidy = block_subsidy(block_height, already_generated);
        return already_generated
            .checked_add(u128::from(subsidy))
            .ok_or_else(|| TxError::new(ErrorCode::BlockErrParse, "already_generated overflow"));
    }
    Ok(already_generated)
}

/// utxo_set_hash computes a deterministic SHA3-256 digest over the UTXO set.
/// Must match Go consensus.UtxoSetHash and rubin-node chainstate for parity.
pub(crate) fn utxo_set_hash(utxos: &HashMap<Outpoint, UtxoEntry>) -> [u8; 32] {
    let mut items: Vec<([u8; 36], &UtxoEntry)> = Vec::with_capacity(utxos.len());
    for (outpoint, entry) in utxos {
        let mut key = [0u8; 36];
        key[..32].copy_from_slice(&outpoint.txid);
        key[32..].copy_from_slice(&outpoint.vout.to_le_bytes());
        items.push((key, entry));
    }
    // sort_unstable_by avoids decorate/sort/undecorate copies of the
    // [u8; 36] key that sort_by_key/sort_unstable_by_key would do —
    // important on the consensus digest path with large UTXO sets.
    #[allow(clippy::unnecessary_sort_by)]
    items.sort_unstable_by(|a, b| a.0.cmp(&b.0));

    let mut buf = Vec::with_capacity(UTXO_SET_HASH_DST.len() + 8 + items.len() * 64);
    buf.extend_from_slice(UTXO_SET_HASH_DST);
    buf.extend_from_slice(&(items.len() as u64).to_le_bytes());

    for (key, entry) in items {
        buf.extend_from_slice(&key);
        buf.extend_from_slice(&entry.value.to_le_bytes());
        buf.extend_from_slice(&entry.covenant_type.to_le_bytes());
        encode_compact_size(entry.covenant_data.len() as u64, &mut buf);
        buf.extend_from_slice(&entry.covenant_data);
        buf.extend_from_slice(&entry.creation_height.to_le_bytes());
        buf.push(u8::from(entry.created_by_coinbase));
    }

    Sha3_256::digest(&buf).into()
}

#[cfg(test)]
#[path = "connect_block_inmem_digest_tests.rs"]
mod tests;
